File: templane-ts/src/type-checker.ts
import { TemplaneFieldType, TypeCheckError, TypedSchema } from './models';

function levenshtein(s1: string, s2: string): number {
  const m = s1.length;
  const n = s2.length;
  const dp = Array.from({ length: n + 1 }, (_, i) => i);
  for (let i = 1; i <= m; i++) {
    let prev = dp[0];
    dp[0] = i;
    for (let j = 1; j <= n; j++) {
      const temp = dp[j];
      dp[j] = s1[i - 1] === s2[j - 1] ? prev : 1 + Math.min(prev, dp[j], dp[j - 1]);
      prev = temp;
    }
  }
  return dp[n];
}

function typeNameOf(value: unknown): string {
  if (typeof value === 'boolean') return 'boolean';
  if (typeof value === 'number') return 'number';
  if (typeof value === 'string') return 'string';
  if (value === null) return 'null';
  if (Array.isArray(value)) return 'list';
  if (typeof value === 'object') return 'object';
  return typeof value;
}

export function check(
  schema: TypedSchema,
  data: Record<string, unknown>,
  prefix: string = '',
): TypeCheckError[] {
  const errors: TypeCheckError[] = [];

  for (const [fieldName, field] of Object.entries(schema.fields)) {
    const path = prefix ? `${prefix}.${fieldName}` : fieldName;
    if (!(fieldName in data)) {
      if (field.required) {
        errors.push({
          code: 'missing_required_field',
          field: path,
          message: `Required field '${path}' is missing`,
        });
      }
    } else {
      errors.push(...checkType(field.type, data[fieldName], path));
    }
  }

  const known = Object.keys(schema.fields);
  for (const key of Object.keys(data)) {
    if (!(key in schema.fields)) {
      const path = prefix ? `${prefix}.${key}` : key;
      let closest: string | null = null;
      let minDist = Infinity;
      for (const k of known) {
        const d = levenshtein(key, k);
        if (d < minDist) {
          minDist = d;
          closest = k;
        }
      }
      if (closest !== null && minDist <= 3) {
        errors.push({
          code: 'did_you_mean',
          field: path,
          message: `Unknown field '${path}'. Did you mean '${closest}'?`,
        });
      } else {
        errors.push({
          code: 'unknown_field',
          field: path,
          message: `Field '${path}' is not defined in schema`,
        });
      }
    }
  }

  return errors;
}

function checkType(fieldType: TemplaneFieldType, value: unknown, path: string): TypeCheckError[] {
  const errors: TypeCheckError[] = [];

  switch (fieldType.kind) {
    case 'string':
      if (typeof value !== 'string') {
        errors.push({
          code: 'type_mismatch',
          field: path,
          message: `Field '${path}' expected string, got ${typeNameOf(value)}`,
        });
      }
      break;
    case 'number':
      if (typeof value !== 'number') {
        errors.push({
          code: 'type_mismatch',
          field: path,
          message: `Field '${path}' expected number, got ${typeNameOf(value)}`,
        });
      }
      break;
    case 'boolean':
      if (typeof value !== 'boolean') {
        errors.push({
          code: 'type_mismatch',
          field: path,
          message: `Field '${path}' expected boolean, got ${typeNameOf(value)}`,
        });
      }
      break;
    case 'null':
      if (value !== null) {
        errors.push({
          code: 'type_mismatch',
          field: path,
          message: `Field '${path}' expected null, got ${typeNameOf(value)}`,
        });
      }
      break;
    case 'enum':
      if (typeof value !== 'string' || !fieldType.values.includes(value)) {
        errors.push({
          code: 'invalid_enum_value',
          field: path,
          message: `Field '${path}' value '${value}' not in enum [${fieldType.values.join(', ')}]`,
        });
      }
      break;
    case 'list':
      if (!Array.isArray(value)) {
        errors.push({
          code: 'type_mismatch',
          field: path,
          message: `Field '${path}' expected list, got ${typeNameOf(value)}`,
        });
      } else {
        value.forEach((item, i) => {
          errors.push(...checkType(fieldType.item_type, item, `${path}[${i}]`));
        });
      }
      break;
    case 'object':
      if (typeof value !== 'object' || value === null || Array.isArray(value)) {
        errors.push({
          code: 'type_mismatch',
          field: path,
          message: `Field '${path}' expected object, got ${typeNameOf(value)}`,
        });
      } else {
        const subSchema: TypedSchema = { id: '', fields: fieldType.fields };
        errors.push(...check(subSchema, value as Record<string, unknown>, path));
      }
      break;
  }

  return errors;
}

File: templane-python/src/templane_core/type_checker.py
from __future__ import annotations
from .models import (
    TypedSchema, TemplaneField, TemplaneFieldType,
    StringType, NumberType, BooleanType, NullType, EnumType, ListType, ObjectType,
    TypeCheckError,
)


def _levenshtein(s1: str, s2: str) -> int:
    m, n = len(s1), len(s2)
    dp = list(range(n + 1))
    for i in range(1, m + 1):
        prev, dp[0] = dp[0], i
        for j in range(1, n + 1):
            temp = dp[j]
            dp[j] = prev if s1[i - 1] == s2[j - 1] else 1 + min(prev, dp[j], dp[j - 1])
            prev = temp
    return dp[n]


def _python_type_name(value: object) -> str:
    if isinstance(value, bool):
        return "boolean"
    if isinstance(value, (int, float)):
        return "number"
    if isinstance(value, str):
        return "string"
    if value is None:
        return "null"
    if isinstance(value, list):
        return "list"
    if isinstance(value, dict):
        return "object"
    return type(value).__name__


def check(schema: TypedSchema, data: dict, _prefix: str = "") -> list[TypeCheckError]:
    errors: list[TypeCheckError] = []

    for field_name, field in schema.fields.items():
        path = f"{_prefix}.{field_name}" if _prefix else field_name
        if field_name not in data:
            if field.required:
                errors.append(TypeCheckError(
                    code="missing_required_field",
                    field=path,
                    message=f"Required field '{path}' is missing",
                ))
        else:
            errors.extend(_check_type(field.type, data[field_name], path, schema, data, _prefix))

    for key in data:
        if key not in schema.fields:
            path = f"{_prefix}.{key}" if _prefix else key
            known = list(schema.fields.keys())
            closest = min(known, key=lambda k: _levenshtein(key, k)) if known else None
            if closest is not None and _levenshtein(key, closest) <= 3:
                errors.append(TypeCheckError(
                    code="did_you_mean",
                    field=path,
                    message=f"Unknown field '{path}'. Did you mean '{closest}'?",
                ))
            else:
                errors.append(TypeCheckError(
                    code="unknown_field",
                    field=path,
                    message=f"Field '{path}' is not defined in schema",
                ))

    return errors


def _check_type(
    field_type: TemplaneFieldType,
    value: object,
    path: str,
    schema: TypedSchema,
    data: dict,
    prefix: str,
) -> list[TypeCheckError]:
    errors: list[TypeCheckError] = []

    if isinstance(field_type, StringType):
        if not isinstance(value, str):
            errors.append(TypeCheckError(
                code="type_mismatch",
                field=path,
                message=f"Field '{path}' expected string, got {_python_type_name(value)}",
            ))

    elif isinstance(field_type, NumberType):
        if isinstance(value, bool) or not isinstance(value, (int, float)):
            errors.append(TypeCheckError(
                code="type_mismatch",
                field=path,
                message=f"Field '{path}' expected number, got {_python_type_name(value)}",
            ))

    elif isinstance(field_type, BooleanType):
        if not isinstance(value, bool):
            errors.append(TypeCheckError(
                code="type_mismatch",
                field=path,
                message=f"Field '{path}' expected boolean, got {_python_type_name(value)}",
            ))

    elif isinstance(field_type, EnumType):
        if value not in field_type.values:
            errors.append(TypeCheckError(
                code="invalid_enum_value",
                field=path,
                message=f"Field '{path}' value '{value}' not in enum [{', '.join(field_type.values)}]",
            ))

    elif isinstance(field_type, ListType):
        if not isinstance(value, list):
            errors.append(TypeCheckError(
                code="type_mismatch",
                field=path,
                message=f"Field '{path}' expected list, got {_python_type_name(value)}",
            ))
        else:
            for i, item in enumerate(value):
                errors.extend(_check_type(field_type.item_type, item, f"{path}[{i}]", schema, data, prefix))

    elif isinstance(field_type, ObjectType):
        if not isinstance(value, dict):
            errors.append(TypeCheckError(
                code="type_mismatch",
                field=path,
                message=f"Field '{path}' expected object, got {_python_type_name(value)}",
            ))
        else:
            sub_schema = TypedSchema(id="", fields=field_type.fields)
            errors.extend(check(sub_schema, value, _prefix=path))

    return errors
